### ISSUE TITLE
Rule S4586: False positive when returning null from inside Task.Run ?

### ISSUE THREAD

**Issue by ruudkobes** [SonarSource/sonar-csharp#1436]
### Description

A `null` returned from a non-async lambda expression, executed by `Task.Run<T>()`, is regarded as non-compliant by the current implementation of rule S4586.

### Actual and expected behaviour

    public Task<object> GetFooAsync()
    {
        return Task.Run(() => null) // Should be compliant, but is regarded as non-compliant
    }

### Known workarounds
One could actually adhere to the rule, which would result in using a different overload of `Task.Run`, which is able to unwrap the returned `Task`

Instead of:

    Task<TResult> Run<TResult> (Func<TResult> function)

one would be using

    Task<TResult> Run<TResult> (Func<Task<TResult>> function)

**Comment by Evangelink** [comment]
Hi @ruudkobes,

Your code sample is invalid and requires an explicit cast to `object` or the return type to be `Task`:

```
public Task<object> GetFooAsync()
{
    return Task.Run(() => (object)null);
}
```

With or without the fix I cannot reproduce the False Positive you are mentioning (testing with the latest release). Which version of SonarLint/SonarAnalyzer.CSharp are you using?

**Comment by ruudkobes** [comment]
Hi @Evangelink,

Sorry about that. Was a bit to quick modifying an existing sample. The sample below more accurately reflects the situation where the report occurs.

This is reported by SonarScanner for MSBuild 4.2, which runs as a build step in VSTS.

```C#
public Task<object> GetFooAsync()
{
    return Task.Run(() =>
    {
        if (true)
        {
            return new object();
        }
        else
        {
            return null;
        }
    });
}
```

**Comment by Evangelink** [comment]
Hi @ruudkobes,

There is indeed a FP in here! Thank you for the feedback! We will make sure to fix that on the next version.
